### PR TITLE
Add Timeout support

### PR DESCRIPTION
**Pull request by jayjamesjay**
#28 
Basic implementation of timeout for entire request. The default timeout is 3 hours.

### CODE REVIEW

File: src/request.rs [hunk 1]
```diff
@@ -1,46 +1,102 @@
...
+const BUF_SIZE: usize = 8 * 1024;
+const SMALL_BUF_SIZE: usize = 8 * 10;
+const TEST_FREQ: usize = 200;
+const CONNECTION_TIMEOUT: u64 = 3 * 3600;
 
-///Copies data from `reader` to `writer` until the specified `val`ue is reached.
+///Copies data from `reader` to `writer` until the `timeout` is reached.
 ///Returns how many bytes has been read.
-pub fn copy_until<R, W>(reader: &mut R, writer: &mut W, val: &[u8]) -> Result<usize, io::Error>
+pub fn copy_with_timeout<R, W>(reader: &mut R, writer: &mut W, timeout: Instant) -> io::Result<u64>
```
Comment (Shnatsel) [inline src/request.rs]: `timeout` here should be called `deadline`. Timeout would be of type `Duration`, while this is an `Instant` by which you need it to be done.

File: src/request.rs [hunk 2]
```diff
@@ -587,6 +676,32 @@ impl<'a> Request<'a> {
         self
     }
 
+    ///Sets connection timeout of request.
+    ///By default timeout is set to 3 hours.
```
Comment (Shnatsel) [inline src/request.rs:711]: There should not be a default timeout for the entire request. That would unconditionally interrupt long downloads.

File: src/request.rs [hunk 3]
```diff
@@ -1,46 +1,102 @@
...
     R: Read + ?Sized,
     W: Write + ?Sized,
 {
-    let mut buf = Vec::with_capacity(200);
+    let mut buf = [0; BUF_SIZE];
+    let mut copied = 0;
+    let mut counter = 0;
+
+    loop {
+        let len = match reader.read(&mut buf) {
```
Comment (Shnatsel) [inline src/request.rs:62]: I'm not 100% sure, but I expect this call would block indefinitely if it's waiting on more data to arrive, defeating the point of the deadline.

I believe you can get around this by setting read timeout to `initial_request_time + timeout - current_time` every once in a while, see e.g. https://github.com/neonmoe/minreq/commit/41ef16e33501712e64a5711a7f4bc132d9bfbe40

File: src/request.rs (hunk 1 above)
Comment (jayjamesjay) [reply src/request.rs]: You're right. I've renamed it

File: src/request.rs (hunk 2 above)
Comment (jayjamesjay) [reply src/request.rs:711]: Now timeout is optional (`Option<Duration>`). By default there is no timeout.

File: src/request.rs (hunk 3 above)
Comment (jayjamesjay) [reply src/request.rs:62]: I need to take a closer look at this.

**Comment by jayjamesjay** [comment]
@Shnatsel I've resolved all the issues you mentioned in the comments, except for the `initial_request_time + timeout - current_time` - I think that just checking once in a while if deadline hasn't been reached should do the job.

Please let me know if you have any other suggestions.

**Pull request merged**